proposals/0000-Move-Platforms-to-Own-Packages.md

```diff
@@ -0,0 +1,273 @@
+
+The native build scripts (gradle, BUCK, xcode, etc.) will all need to be updated for the new code locations. Great care must be taken to not break the existing large ecosystem of native compnents on NPM!
+
+## Drawbacks
+
+ - A lot of old open pull requests will need to be rebased if the codebase is re-arranged
+ - The `react-native-ios` and `react-native-android` packages on NPM already have the following versions on them:
+   > - `react-native-ios`: `1.0.0`, `1.0.1`, `1.1.0`
+   > - `react-native-android`: `1.0.0`, `2.0.0`, `3.0.0`
+   
+   This will make React Native's future semver bump to `1.0.0` awkward, if not impossible. Unless it does what React did and jumps to version `60.0.0` or something like that. Or starts at version `4.0.0`, which seems kinda random. NPM also says that for various technical and security reasons, these versions can not be deleted and re-used. I've `npm deprecate`d them, which is the best I can do sadly.
+ - While community modules that want to cover multiple versions of RN could use the "bridge" versions of the `{ios,android}` modules to ease the transition there are packages that are not maintained that would get broken by a split like this. This is also an issue for Slimmened modules as well.
+ - Even with a codemod to ease the transition, I don't imagine that transitioning codebases over will be something that is trivial. Maybe more than 2 versions should be given as a grace period before everything moves over?
+ - Old documentation and tutorials will need to be updated.
+
+## Alternatives
+
+I believe that this is the best way forward for helping improve React Native's third party platform support, and also for helping improve the existing in-tree platforms by organizing platform-specific things into their own packages. If it's not, hopefully this will spawn discussions on better alternatives.
+
+## Adoption strategy
+
+There could be "bridge" versions of the `react-native-{ios,android}` modules for pre-split versions of React Native that simply import platform specific stuff from `react-native`. This would make it easier for people to make React Native addons that cover wider ranges of RN versions
+
+For a release or so, `react-native-{ios,android}` should probably be thin modules that just re-import stuff from `react-native` before the code actually gets moved over, to help people transition.
+
+For example, the import for `DatePickerIOS` in `react-native` would look like:
+
+```js
+  get DatePickerIOS() {
+    warnMoved('react-native-ios', 'DatePickerIOS');
+    return require('DatePickerIOS');
+  },
+```
+
+(where `warnMoved` is a hypothetical function warning once that an api is moved to another module)
+
+And `react-native-ios` would use Haste to import the module from `react-native`:
```

Comment:
Yeah, that would definitely also make using other packagers (like parcel and webpack) less complex too. @gaearon removed haste from react a while ago and that could be looked to as an example of what could be done:

https://github.com/facebook/react/pull/11303

Platform-specific overrides, however, happen to be really good at doing the platform-specific overrides across multiple packages. A good replacement solution for this would need to be devised. Definitely a subject worthy of a proposal/discussion of its own probably!